### ISSUE TITLE
`checkpoints.txt` not checked when restarting

### ISSUE THREAD

**Issue by PrometheusPi** [ComputationalRadiationPhysics/picongpu#4454]
With openPMD-api automated series read in, only the existence of an (ADIOS) checkpoint file is enough to try a restart from that file, even if it is corrupted due to a failed writing of the checkpoint.
Thus, restarting currently requires manual cleanup. 
Using the last entry in `checkpoints.txt` could avoid that problem and enable automatic restarts.

**Comment by psychocoderHPC** [comment]
To restart form a checkpoint all plugin checkpoints must be valid too.
To handle invalid restart files we need
  - extend the plugin interface with a method where each plugin is validating there checkpoint file
  - openPMD files need a way report to the openPMD-api user that a file is valid.

**Comment by PrometheusPi** [comment]
@franzpoeschel is there are way for openPMD-api to report corrupted iterations in a series?

**Comment by franzpoeschel** [comment]
The upcoming openPMD-api release (see https://github.com/openPMD/openPMD-api/pull/1391) will recover from errors at parsing time. An iteration that is entirely unreadable will be skipped.
If however only some single records or so are broken, then only those will skipped and the rest of the iteration will be shown (might happen in HDF5).
Note that this covers not yet all kinds of errors that might happen at read time.

**Comment by psychocoderHPC** [comment]
@PrometheusPi Please keep in mind that every plugin can write checkpoint data, so it is not enough to check the openPMD files only.
@franzpoeschel IMO the more important question is: Is there a call in openPMD where I can pass in a file and ask if this file is guaranteed fully written? Where fully written means also all disk IO is flushed and writing was not canceled during the flush.
Does openPMD files contain an internal flag to mark that all series are successfully flushed which is written after the flush?

**Comment by franzpoeschel** [comment]
No, openPMD does not explicitly have any atomic writing functionality (yet)

**Comment by PrometheusPi** [comment]
@franzpoeschel I do not get, why an atomic writing needs to exist for that, could you elaborate?

**Comment by franzpoeschel** [comment]
Doing this would be an atomic write, it's how such things are commonly implemented. Use a flag to indicate if the iteration is there or not. It's either fully there or not at all.

**Comment by PrometheusPi** [comment]
@franzpoeschel thanks for the explanation. I naively thought an array of length equal number of writers that is set to a specify (True) value by each writer after they are finished with everything else would be sufficient.

**Comment by ax3l** [comment]
>  Does openPMD files contain an internal flag to mark that all series are successfully flushed which is written after the flush?

Originally, the idea was to write the `/openPMD` attribute last to signal the writing was complete.

At the same time, the request for an atomic completion flag collides in practice with other user requests, which ask to "try to recover and read as much as possible" of a crashed simulation by default.

**Comment by franzpoeschel** [comment]
> At the same time, the request for an atomic completion flag collides in practice with other user requests, which ask to "try to recover and read as much as possible" of a crashed simulation by default.

In practise, can't this be solved by having a call `auto Iteration::finalized() -> bool` that evaluates such a flag, but the Iteration is still parsed and presented to the user anyway?

**Comment by ax3l** [comment]
That would work.

I am not sure we get such a guarantee from HDF5 or ADIOS2 though:
> IMO the more important question is: Is there a call in openPMD where I can pass in a file and ask if this file is guaranteed fully written? Where fully written means also all disk IO is flushed and writing was not canceled during the flush.

We can try somewhat and rely that closed write steps imply things are on the way to disk ... and hope FS-worlds and MPI-world are magically in sync (they usually are not).

**Comment by ax3l** [comment]
I think the underlying resolution to this issue in PIConGPU is to delay writing the entry in `checkpoint.txt` to when all checkpoints from plugins & openPMD-api are written.

Recover only from full checkpoints. Trying to "fix" a corrupted (partial) checkpoint indeed needs manual intervention - it should by default not be taken to avoid surprises and undefined recovery of sim state.

Did I get this right or is there more context to the issue that I missed in the description? From the issue description, it feels to me there is a specific sim crash situation here as a premise that I don't understand.

**Comment by franzpoeschel** [comment]
Doing this somewhat securely at write time would probably mean

1. In HDF5, at least one additional `MPI_Barrier()` / one additional parallel write.
2. In ADIOS2, the attribute would only need to be set right before `EndStep()` / `Engine::Close()`, since ADIOS2 operations are somewhat atomic by themselves already.
    Currently, ADIOS2 only shows such attribute after successfully completing such a call anyway, but this might prepare for a situation when `PerformDataWrite()` creates actually legible data.

Due to additional performance overhead, such a feature should be opt-in.

**Comment by ax3l** [comment]
I am not sure we address the right issue here, let's wait for a bit more context from @PrometheusPi .

For the points, the following challenges:
1. HDF5: MPI_Barrier != FS sync. There is no guarantee in MPI-I/O with regards to "landed on the FS". We can add an attribute for iteration finalize, flush/close, and then MPI barrier, but it's mostly a guess.
2. ADIOS2 steps should actually not be visible to ADIOS2 readers at all before EndStep was called in the writer, according to Norbert. With steps, they are transactional when we try to read them. I agree, same attribute can still be written last, e.g., for iteration finalize.

**Comment by franzpoeschel** [comment]
> I am not sure we address the right issue here, let's wait for a bit more context from @PrometheusPi .

True :)

>     1. HDF5: MPI_Barrier != FS sync. There is no guarantee in MPI-I/O with regards to "landed on the FS". We can hope and add an attribute, but it's mostly a guess.

Shouldn't one parallel write at least precede the other? ("Parallel write the last data in an iteration -> parallel write the flag")
So, if the flag is there in the dataset, the other previous writes must have been successful too?
You know the behavior of HDF5 better than I do here.
 
>     2. ADIOS2 steps should actually not be visible at all before EndStep, according to Norbert. They are transactional when we try to read them.

Yep, currently implementing this is trivial.
But you are currently campaigning Norbert to have this changed, so we should think forward ;)

**Comment by PrometheusPi** [comment]
> I think the underlying resolution to this issue in PIConGPU is to delay writing the entry in `checkpoint.txt` to when all checkpoints from plugins & openPMD-api are written.

I totally agree. If this is the case, not finding the last iteration in  `checkpoints.txt` signals, whether the last output was not correctly written.  

> Recover only from full checkpoints. Trying to "fix" a corrupted (partial) checkpoint indeed needs manual intervention - it should by default not be taken to avoid surprises and undefined recovery of sim state.

I agree. For me it would be fine, that openPMD series just warns that not all files were readable. I would avoid automatic recover and would favor automatic exclusion from the series when opening with read only. 

> Did I get this right or is there more context to the issue that I missed in the description? From the issue description, it feels to me there is a specific sim crash situation here as a premise that I don't understand.

I can not comment on how this is best implemented in hdf5 or adios.
